src/jsonschema.rs

```diff
@@ -0,0 +1,219 @@
+use serde_json::Value;
+use std::collections::HashMap;
+
+/// The type enumeration does not contain any data and is used to determine
+/// available fields in the flattened tag. In JSONSchema parlance, these are
+/// known as `simpleTypes`.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+enum Type {
+    Null,
+    Boolean,
+    Number,
+    Integer,
+    String,
+    Object,
+    Array,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+enum AdditionalProperties {
+    True,
+    False,
+    Object(Box<Tag>),
+}
+
+#[derive(Serialize, Deserialize, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+struct Object {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    properties: Option<HashMap<String, Box<Tag>>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    additional_properties: Option<AdditionalProperties>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pattern_properties: Option<Box<Tag>>,
+}
+
+/// Represent an array of subschemas. This is also known as a `schemaArray`.
+type TagArray = Vec<Box<Tag>>;
+type OneOf = TagArray;
+type AllOf = TagArray;
+
+#[derive(Serialize, Deserialize, Debug, Default)]
+struct Array {
+    // Using Option<TagArray> would support tuple validation
+    #[serde(skip_serializing_if = "Option::is_none")]
+    items: Option<Box<Tag>>,
+}
+
+/// Container for the main body of the schema.
+#[derive(Serialize, Deserialize, Debug, Default)]
+#[serde(rename_all = "camelCase", tag = "type")]
+struct Tag {
+    #[serde(rename = "type", default)]
+    data_type: Value,
+    #[serde(flatten)]
+    object: Object,
+    #[serde(flatten)]
+    array: Array,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    one_of: Option<OneOf>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    all_of: Option<AllOf>,
+    #[serde(flatten)]
+    extra: Option<HashMap<String, Value>>,
+}
+
+struct JSONSchema {
+    data: Tag,
+}
+
+impl JSONSchema {
+    pub fn from_value(value: Value) -> Self {
+        unimplemented!()
+    }
+}
+
+// TODO: impl Into<ast::AST> for JSONSchema
+
+use serde_json::json;
+
+#[test]
+fn test_serialize_type_null() {
+    let schema = Tag {
+        data_type: json!("integer"),
+        ..Default::default()
```

Comment:
I think this is interesting syntax for not having to specify every field. The structs need to derive from the `Default` trait to use this.